README.md
# Odango

[![Build Status](https://travis-ci.org/Asuforce/odango.svg?branch=master)](https://travis-ci.org/Asuforce/odango)
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](LICENSE)

Odango is server software to deploy.

![Odango](/doc/img/odango.jpg)

## Description

Odango is a server for deploying archive files like tarball. Deployment can be automated by combining CI with object storage (S3 compatible).  
Upload the file from CI to obejct storage and odango will handle it for you.

![Architecture diagram](/doc/img/architecture.png)

CI task is only ①. Odango handles ② to ④.

## Requirements

- CI or something to uplad tarball for S3 (or S3 compatible object storage)
- Create Odango server (LINUX recommended)
- Allow Odango server to SSH to Deployment target

## Usage

### Configuration

Edit `~/.odango`

```conf
[server]
endpoint = "deploy"         # Optional: Odango endpoint. Default /deploy
port     = 8080             # Optional: Odango port. Default 8080

[credential]
access_key          = ""    # Required: S3 access key
secret_key          = ""    # Required: S3 secret key
endpoint            = ""    # Optional: You should specify when you use object storage(S3 compatible) without S3
region              = ""    # Required: S3 region
disable_ssl         = false # Optional: Default false
s3_force_path_style = true  # Optional: Default false

[bucket]
name      = ""              # Required: Bucket name
path      = ""              # Optional: You should specify path when your file locate in directry
extension = ""              # Optional: Deploy files extension. Default .tar.gz (Only support .tar.gz now)

[ssh]
hosts     = ["", ""]        # Required: Deployment target servers list
user_name = ""              # Optional: Default $USER
key_path  = ""              # Optional: Default $HOME/.ssh/id_rsa
port      = 22              # Optional: Default 22

[deploy]
dest_dir    = ""            # Required: Specify destination dir in target servers
archive_dir = ""            # Optional: Default /tmp/odango
```

### Run

It is simple to run.

```sh
# Run odango
$ odango
```

#### Unitfile

Create `/etc/systemd/system/odango.service` like below.

```service
[Unit]
Description = Odango Server

[Service]
ExecStart = /usr/local/bin/odango
ExecStop = /bin/kill -HUP $MAINPID
ExecReload = /bin/kill -HUP $MAINPID && /usr/local/bin/odango
Restart = no
Type = simple
User=<username>
Group=<username>

[Install]
WantedBy = multi-user.target
```

## Install

### Linux

```sh
$ curl -sL https://github.com/Asuforce/odango/releases/download/v0.0.1/odango_v0.0.1_linux_amd64.tar.gz |
  sudo tar xz \
  -C /usr/local/bin \
  --strip=1 '*/odango' \
  --no-same-owner \
  --no-same-permissions
```

### Mac

```sh
$ curl -sL -o odango.zip https://github.com/Asuforce/odango/releases/download/v0.0.1/odango_v0.0.1_darwin_amd64.zip
$ unzip odango.zip
$ sudo mv odango/odango /usr/local/bin
```

### Windows

TBD

## Contribution

1. Fork ([https://github.com/Asuforce/odango/fork](https://github.com/Asuforce/odango/fork))
1. Create a feature branch
1. Commit your changes
1. Rebase your local changes against the master branch
1. Run test suite with the `make test` command and confirm that it passes
1. Run `make fmt`
1. Create new Pull Request

## License

This software is released under the MIT License, see [LICENSE](https://github.com/Asuforce/odango/blob/master/LICENSE)

## Author

[Asuforce](https://github.com/Asuforce)


CHANGELOG.md
# Changelog

## [v0.0.2](https://github.com/Asuforce/odango/compare/v0.0.1...v0.0.2) (2019-08-13)

* Compress [#14](https://github.com/Asuforce/odango/pull/14) ([Asuforce](https://github.com/Asuforce))
* Update README and add LICENSE [#12](https://github.com/Asuforce/odango/pull/12) ([Asuforce](https://github.com/Asuforce))
* Add validation [#13](https://github.com/Asuforce/odango/pull/13) ([Asuforce](https://github.com/Asuforce))
* Update configfile [#11](https://github.com/Asuforce/odango/pull/11) ([Asuforce](https://github.com/Asuforce))
* Fix to read config file [#8](https://github.com/Asuforce/odango/pull/8) ([Asuforce](https://github.com/Asuforce))
* Add Changelog [#7](https://github.com/Asuforce/odango/pull/7) ([Asuforce](https://github.com/Asuforce))

## [v0.0.1](https://github.com/Asuforce/odango/compare/8a3bd9090cf8...v0.0.1) (2019-07-25)

* Fix typo [#6](https://github.com/Asuforce/odango/pull/6) ([Asuforce](https://github.com/Asuforce))
* Create release flow [#5](https://github.com/Asuforce/odango/pull/5) ([Asuforce](https://github.com/Asuforce))
* Change name to odango [#4](https://github.com/Asuforce/odango/pull/4) ([Asuforce](https://github.com/Asuforce))
* Change name [#3](https://github.com/Asuforce/odango/pull/3) ([Asuforce](https://github.com/Asuforce))
* Rename to gorigin [#2](https://github.com/Asuforce/odango/pull/2) ([Asuforce](https://github.com/Asuforce))
* Make gongcha [#1](https://github.com/Asuforce/odango/pull/1) ([Asuforce](https://github.com/Asuforce))


config/config.go
package config

import (
	"fmt"
	"log"
	"os"
	"reflect"
	"strings"

	"github.com/BurntSushi/toml"
)

type (
	// Config is odango configuration structure
	Config struct {
		HomeDir    string
		Server     Server
		Credential Credential
		Bucket     Bucket
		SSH        SSH `toml:"ssh"`
		Deploy     Deploy
	}

	// Server is server configuration
	Server struct {
		Endpoint string
		Port     int
	}

	// Credential is object storage credential
	Credential struct {
		AccessKey        string `toml:"access_key"`
		SecretKey        string `toml:"secret_key"`
		Endpoint         string
		Region           string
		DisableSSL       bool `toml:"disable_ssl"`
		S3ForcePathStyle bool `toml:"s3_force_path_style"`
	}

	// Bucket is object storage information
	Bucket struct {
		Name      string
		Path      string
		Extension string
	}

	// SSH is ssh configuration
	SSH struct {
		UserName string `toml:"user_name"`
		KeyPath  string `toml:"key_path"`
		Hosts    []string
		Port     int
	}

	// Deploy is deploy configuration
	Deploy struct {
		ArchiveDir string `toml:"archive_dir"`
		DestDir    string `toml:"dest_dir"`
	}
)

// Read is checking and reading odango configuration file
func (c *Config) Read() error {
	if !isExist(c.HomeDir + "/.odango") {
		if err := c.createFile(); err != nil {
			return err
		}
	}

	if _, err := toml.DecodeFile(c.HomeDir+"/.odango", &c); err != nil {
		log.Fatalf("Unable to credential file, %v", err)
	}

	c.validateEndpoint()
	c.validateAccessKey()
	c.validate()
	c.checkFormat()

	return nil
}

func isExist(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func (c *Config) createFile() error {
	config := `[server]
#endpoint = "deploy"
#port     = 8080

[credential]
access_key           = ""
secret_key           = ""
#endpoint            = ""
region 	             = ""
#disable_ssl         = false
#s3_force_path_style = true

[bucket]
name       = ""
path       = ""
#extension = ".tar.gz"

[ssh]
hosts     = ["", ""]
#user_name = ""      # Default: $USER
#key_path  = ""      # Default: $HOME/.ssh/id_rsa
#port      = 22

[deploy]
dest_dir    = ""
#archive_dir = "/tmp/odango"
`

	file, err := os.Create(c.HomeDir + "/.odango")
	if err != nil {
		return err
	}
	defer file.Close()
	fmt.Fprint(file, config)

	return nil
}

func (c *Config) checkFormat() {
	c.Bucket.Path = formatPath(c.Bucket.Path)
	c.Deploy.ArchiveDir = formatPath(c.Deploy.ArchiveDir)
	c.Deploy.DestDir = formatPath(c.Deploy.DestDir)
}

func (c *Config) validate() {
	if isZero(c.Credential.SecretKey) {
		log.Fatal("Please set the credential secret_key")
	}
	if isZero(c.Credential.Region) {
		log.Fatal("Please set the credential region")
	}
	if isZero(c.Bucket.Name) {
		log.Fatal("Please set the bucket region")
	}
	if isZero(c.SSH.UserName) {
		c.SSH.UserName = os.Getenv("USER")
	}
	if isZero(c.SSH.KeyPath) {
		c.SSH.KeyPath = os.Getenv("HOME") + "/.ssh/id_rsa"
	}
	if isZero(c.SSH.Hosts) {
		log.Fatal("Please set the deploy target hostname")
	}
	if isZero(c.SSH.Port) {
		c.SSH.Port = 22
	}
	if isZero(c.Deploy.ArchiveDir) {
		c.Deploy.ArchiveDir = "/tmp/odango"
	}
	if isZero(c.Deploy.DestDir) {
		log.Fatal("Please set the deploy dest_dir")
	}
}

func (c *Config) validateEndpoint() {
	if isZero(c.Server.Endpoint) {
		c.Server.Endpoint = "/deploy/"
		return
	}
	c.Server.Endpoint = formatPath(c.Server.Endpoint)
}

func (c *Config) validatePort() {
	if isZero(c.Server.Port) {
		c.Server.Port = 8080
		return
	}
}

func (c *Config) validateAccessKey() {
	if isZero(c.Credential.AccessKey) {
		log.Fatal("Please set the credential access_key")
	}
}

func isZero(x interface{}) bool {
	return reflect.DeepEqual(x, reflect.Zero(reflect.TypeOf(x)).Interface())
}

func formatPath(s string) string {
	if !strings.HasPrefix(s, "/") {
		s = "/" + s
	}
	if !strings.HasSuffix(s, "/") {
		s = s + "/"
	}
	return s
}
